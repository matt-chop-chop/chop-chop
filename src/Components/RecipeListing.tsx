import { Box, Flex, Heading, Link } from "@chakra-ui/react";
import NextLink from "next/link";
import { RecipeDetailsGrid, RecipeImage, RecipeTags } from ".";

type RecipeListingProps = {
  recipe: Recipe;
};

const RecipeListing = ({ recipe }: RecipeListingProps) => {
  return (
    <Link
      as={NextLink}
      href={`/recipe/${recipe.id}`}
      _hover={{ textDecoration: "none" }}
    >
      <Flex
        background="var(--chakra-colors-light-background)"
        border="2px solid var(--chakra-colors-light-text)"
        borderRadius={4}
        flexDirection="column"
        height="100%"
        justifyContent="space-between"
        minHeight="300px"
        p={4}
        transition="transform 250ms ease-out"
        width="100%"
        _hover={{
          border: "2px solid var(--chakra-colors-light-emphasis)",
          cursor: "pointer",
          transition: "transform 250ms ease-in",
          transform: "scale(1.01)",
        }}
      >
        <Box>
          <RecipeImage
            image={recipe.thumbnail}
            maxHeight={["200px", "200px", "200px"]}
            defaultLogoWidth={["60%", "70%", "55%"]}
          />
          <Heading
            as="h3"
            textDecoration="underline"
            mt={4}
            variant="heading-small"
          >
            {recipe.name}
          </Heading>
          <RecipeDetailsGrid
            area={recipe.area}
            category={recipe.category}
            ingredientCount={recipe.ingredientCount}
            time={recipe.time}
            gridTemplateColumns={[
              "repeat(2, 1fr)",
              "repeat(2, 1fr)",
              "repeat(2, 1fr)",
            ]}
          />
        </Box>
        <RecipeTags tags={recipe.tags} />
      </Flex>
    </Link>
  );
};

export default RecipeListing;
