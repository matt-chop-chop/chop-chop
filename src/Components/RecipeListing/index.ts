export { default as RecipeListing } from "./RecipeListing";
