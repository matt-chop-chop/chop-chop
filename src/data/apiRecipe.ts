export const apiRecipe: ApiRecipe = {
  idMeal: "52922",
  strMeal: "Prawn & Fennel Bisque",
  strDrinkAlternate: null,
  strCategory: "Side",
  strArea: "French",
  strInstructions:
    "Shell the prawns, then fry the shells in the oil in a large pan for about 5 mins. Add the onion, fennel and carrots and cook for about 10 mins until the veg start to soften. Pour in the wine and brandy, bubble hard for about 1 min to drive off the alcohol, then add the tomatoes, stock and paprika. Cover and simmer for 30 mins. Meanwhile, chop the prawns.\r\nBlitz the soup as finely as you can with a stick blender or food processor, then press through a sieve into a bowl. Spend a bit of time really working the mixture through the sieve as this will give the soup its velvety texture.\r\nTip back into a clean pan, add the prawns and cook for 10 mins, then blitz again until smooth. You can make and chill this a day ahead or freeze it for 1 month. Thaw ovenight in the fridge. To serve, gently reheat in a pan with the cream. If garnishing, cook the 8 prawns in a little butter. Spoon into small bowls and top with the prawns and snipped fennel fronds.",
  strMealThumb:
    "https://www.themealdb.com/images/media/meals/rtwwvv1511799504.jpg",
  strTags: "Soup,Warm,Seafood,Shellfish",
  strYoutube: "https://www.youtube.com/watch?v=4thpMbDakgM",
  strIngredient1: "Tiger Prawns",
  strIngredient2: "Olive Oil",
  strIngredient3: "Onion",
  strIngredient4: "Fennel",
  strIngredient5: "Carrots",
  strIngredient6: "Dry White Wine",
  strIngredient7: "Brandy",
  strIngredient8: "Chopped Tomatoes",
  strIngredient9: "Fish Stock",
  strIngredient10: "Paprika",
  strIngredient11: "Double Cream",
  strIngredient12: "Prawns",
  strIngredient13: "",
  strIngredient14: "",
  strIngredient15: "",
  strIngredient16: "",
  strIngredient17: "",
  strIngredient18: "",
  strIngredient19: "",
  strIngredient20: "",
  strMeasure1: "450g",
  strMeasure2: "4 tbs",
  strMeasure3: "1 large",
  strMeasure4: "1 large",
  strMeasure5: "2 chopped",
  strMeasure6: "150ml",
  strMeasure7: "1 tbs",
  strMeasure8: "400g",
  strMeasure9: "1L",
  strMeasure10: "2 pinches",
  strMeasure11: "150ml",
  strMeasure12: "8",
  strMeasure13: "",
  strMeasure14: "",
  strMeasure15: "",
  strMeasure16: "",
  strMeasure17: "",
  strMeasure18: "",
  strMeasure19: "",
  strMeasure20: "",
  strSource:
    "https://www.bbcgoodfood.com/recipes/71600/prawn-and-fennel-bisque",
  strImageSource: null,
  strCreativeCommonsConfirmed: null,
  dateModified: null,
};
