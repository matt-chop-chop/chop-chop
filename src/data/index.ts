export { recipe } from "./recipe";
export { apiRecipe } from "./apiRecipe";
