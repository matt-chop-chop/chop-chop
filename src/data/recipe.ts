export const recipe: Recipe = {
  id: "52922",
  area: "French",
  category: "Side",
  image: "https://www.themealdb.com/images/media/meals/rtwwvv1511799504.jpg",
  ingredients: [
    {
      name: "Tiger Prawns",
      measurement: "450g",
      image: "https://www.themealdb.com/images/ingredients/Tiger Prawns.png",
    },
    {
      name: "Olive Oil",
      measurement: "4 tbs",
      image: "https://www.themealdb.com/images/ingredients/Olive Oil.png",
    },
    {
      name: "Onion",
      measurement: "1 large",
      image: "https://www.themealdb.com/images/ingredients/Onion.png",
    },
    {
      name: "Fennel",
      measurement: "1 large",
      image: "https://www.themealdb.com/images/ingredients/Fennel.png",
    },
    {
      name: "Carrots",
      measurement: "2 chopped",
      image: "https://www.themealdb.com/images/ingredients/Carrots.png",
    },
    {
      name: "Dry White Wine",
      measurement: "150ml",
      image: "https://www.themealdb.com/images/ingredients/Dry White Wine.png",
    },
    {
      name: "Brandy",
      measurement: "1 tbs",
      image: "https://www.themealdb.com/images/ingredients/Brandy.png",
    },
    {
      name: "Chopped Tomatoes",
      measurement: "400g",
      image:
        "https://www.themealdb.com/images/ingredients/Chopped Tomatoes.png",
    },
    {
      name: "Fish Stock",
      measurement: "1L",
      image: "https://www.themealdb.com/images/ingredients/Fish Stock.png",
    },
    {
      name: "Paprika",
      measurement: "2 pinches",
      image: "https://www.themealdb.com/images/ingredients/Paprika.png",
    },
    {
      name: "Double Cream",
      measurement: "150ml",
      image: "https://www.themealdb.com/images/ingredients/Double Cream.png",
    },
    {
      name: "Prawns",
      measurement: "8",
      image: "https://www.themealdb.com/images/ingredients/Prawns.png",
    },
  ],
  instructions: [
    "Shell the prawns, then fry the shells in the oil in a large pan for about 5 mins. ",
    "Add the onion, fennel and carrots and cook for about 10 mins until the veg start to soften. ",
    "Pour in the wine and brandy, bubble hard for about 1 min to drive off the alcohol, then add the tomatoes, stock and paprika. ",
    "Cover and simmer for 30 mins. ",
    "Meanwhile, chop the prawns.\r\nBlitz the soup as finely as you can with a stick blender or food processor, then press through a sieve into a bowl. ",
    "Spend a bit of time really working the mixture through the sieve as this will give the soup its velvety texture.\r\nTip back into a clean pan, add the prawns and cook for 10 mins, then blitz again until smooth. ",
    "You can make and chill this a day ahead or freeze it for 1 month. ",
    "Thaw ovenight in the fridge. ",
    "To serve, gently reheat in a pan with the cream. ",
    "If garnishing, cook the 8 prawns in a little butter. ",
    "Spoon into small bowls and top with the prawns and snipped fennel fronds.",
  ],
  name: "Prawn & Fennel Bisque",
  thumbnail:
    "https://www.themealdb.com/images/media/meals/rtwwvv1511799504.jpg/preview",
  source: "https://www.bbcgoodfood.com/recipes/71600/prawn-and-fennel-bisque",
  tags: ["Soup", "Warm", "Seafood", "Shellfish"],
  ingredientCount: 12,
  time: 75,
  youtube: "https://www.youtube.com/watch?v=4thpMbDakgM",
};
