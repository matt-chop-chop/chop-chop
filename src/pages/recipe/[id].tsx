import { ErrorCard, LoadingCard, MethodStep, Toast } from "@/Components";
import { useRecipe } from "@/hooks";
import { DownloadIcon, EmailIcon, TimeIcon } from "@chakra-ui/icons";
import {
  AspectRatio,
  Box,
  Flex,
  Grid,
  Heading,
  Image,
  Link,
  Tag,
  Text,
  ToastProps,
  useToast,
} from "@chakra-ui/react";
import { useEffect, useState } from "react";
import NextLink from "next/link";
import { fromatCookingTime } from "@/utils";

const parseRecipe = (recipe: Recipe): string => {
  let response =
    recipe.name +
    "\n" +
    `Cooking Time: ${recipe.time}` +
    "\n" +
    `Ingredient Count: ${recipe.ingredientCount}` +
    "\n" +
    `Meal Type: ${recipe.category}` +
    "\n" +
    `Place of Origin: ${recipe.area}` +
    "\n" +
    "Tags:";
  recipe.tags.map((tag) => response + " " + tag);
  response += "\n" + "Ingredients:";
  recipe.ingredients.map(
    (ingredient) =>
      (response += "\n" + `• ${ingredient.measurement} ${ingredient.name}`)
  );
  response += "\n" + "Method:";
  recipe.instructions.map(
    (instruction) => (response += "\n" + `- ${instruction}`)
  );
  response +=
    "\n" +
    `Youtube Tutorial: ${recipe.youtube}` +
    "\n" +
    `Recipe Source: ${recipe.source}`;
  return response;
};

const download = (filename: string, recipe: Recipe) => {
  const link = document.createElement("a");
  link.setAttribute(
    "href",
    "data:text/plain;charset=utf-8," + encodeURIComponent(parseRecipe(recipe))
  );
  link.setAttribute("download", filename);

  link.style.display = "none";
  document.body.appendChild(link);

  link.click();

  document.body.removeChild(link);
};

const Recipe = () => {
  const { recipe, loading, error } = useRecipe();
  const [checkboxes, setCheckboxes] = useState<boolean[]>([]);

  const toast = useToast();

  useEffect(() => {
    if (!loading && recipe && checkboxes.length === 0) {
      setCheckboxes(new Array(recipe.instructions.length).fill(false));
    }
  }, [checkboxes, loading, recipe]);

  if (loading) {
    return <LoadingCard />;
  }

  if (error) {
    return <ErrorCard error={error} />;
  }

  if (!recipe) {
    return <ErrorCard error={new Error("Recipe could not be found")} />;
  }

  const checkRecipeComplete = (updateCheckboxes: boolean[]) => {
    if (
      updateCheckboxes.every((checkbox) => checkbox === true) &&
      updateCheckboxes.length > 0
    ) {
      toast({
        status: "success",
        isClosable: true,
        position: "top-right",
        render: (props: ToastProps) => (
          <Toast
            onClose={props.onClose}
            text="Congratulations, you completed the recipe! Enjoy!"
          />
        ),
      });
    }
  };

  const updateCheckboxes = (checked: boolean, index: number) => {
    const updatedCheckboxes = checkboxes.map((checkbox, i) => {
      if (i === index) return checked;
      else return checkbox;
    });
    setCheckboxes(updatedCheckboxes);

    checkRecipeComplete(updatedCheckboxes);
  };

  return (
    <Flex flexDirection="column">
      <Flex
        borderBottom="2px solid var(--chakra-colors-light-emphasis)"
        flexDirection="column"
        p={4}
      >
        {recipe?.thumbnail ? (
          <Image
            alt="a recipe listing image"
            maxHeight={["200px", "300px", "400px"]}
            objectFit="cover"
            src={recipe.image}
            width="100%"
          />
        ) : (
          <Flex
            alignItems="center"
            background="var(--chakra-colors-light-background)"
            height={["200px", "300px", "400px"]}
            justifyContent="center"
            objectFit="cover"
            p={1}
            width="100%"
          >
            <Image alt="Chop Chop Logo" height="75%" src="/images/logo.png" />
          </Flex>
        )}
        <Flex alignItems="center" justifyContent="space-between" width="100%">
          <Heading
            as="h1"
            textDecoration="underline"
            mt={6}
            variant="heading-large"
          >
            {recipe.name}
          </Heading>
          <DownloadIcon
            boxSize={5}
            cursor="pointer"
            onClick={() => download(`${recipe.name}.txt`, recipe)}
          />
        </Flex>

        <Grid
          columnGap={[8, 8, 6]}
          gridAutoRows="1fr"
          gridTemplateColumns={[
            "repeat(2, 1fr)",
            "repeat(2, 1fr)",
            "repeat(4, 1fr)",
          ]}
          mt={8}
          rowGap={8}
        >
          <Flex flexDirection="column">
            <Flex alignItems="center">
              <Text variant="body-bold">Cooking Time</Text> <TimeIcon ml={2} />
            </Flex>
            <Text mt={1} variant="body">
              {fromatCookingTime(recipe.time)}
            </Text>
          </Flex>
          <Flex flexDirection="column">
            <Text variant="body-bold">Ingredient Count</Text>
            <Text mt={1} variant="body">
              {recipe.ingredientCount}
            </Text>
          </Flex>
          <Flex flexDirection="column">
            <Text variant="body-bold">Meal Type</Text>
            <Text mt={1} variant="body">
              {recipe.category}
            </Text>
          </Flex>
          <Flex flexDirection="column">
            <Text variant="body-bold">Place of Origin</Text>
            <Text mt={1} variant="body">
              {recipe.area}
            </Text>
          </Flex>
        </Grid>
        {recipe.tags.length > 0 && (
          <Flex mt={1}>
            {recipe.tags.map((tag: string, index: number) => (
              <Tag key={index}>{tag}</Tag>
            ))}
          </Flex>
        )}
      </Flex>
      <Flex flexDirection="column" p={4}>
        <Heading as="h2" mb={4} mt={6} variant="heading-medium">
          Ingredients
        </Heading>
        {recipe.ingredients.map((ingredient, index) => (
          <Flex key={index} width="100%">
            <Text ml={[4, 6, 8]} mt={2} variant="body-large">
              • {ingredient.measurement} {ingredient.name}
            </Text>
            {ingredient.image && (
              <Image
                alt="an ingredient image"
                maxHeight="50px"
                ml={4}
                src={ingredient.image}
              />
            )}
          </Flex>
        ))}
        <Heading as="h2" mb={4} mt={6} variant="heading-medium">
          Method
        </Heading>
        {recipe.instructions.map((instruction, index) => (
          <MethodStep
            key={index}
            step={instruction}
            checked={checkboxes[index]}
            setChecked={(checked) => updateCheckboxes(checked, index)}
          />
        ))}
        {recipe.youtube.length > 0 && (
          <Box border="1px solid var(--chakra-colors-light-emphasis)" mt="80px">
            <AspectRatio maxW="100%" ratio={16 / 9}>
              <iframe
                allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
                allowFullScreen
                src={recipe.youtube.replace("watch?v=", "embed/")}
                title="YouTube video player"
              ></iframe>
            </AspectRatio>
          </Box>
        )}
        <Link as={NextLink} href={recipe.source} isExternal mt={4}>
          Click to see the original recipe source
        </Link>
      </Flex>
    </Flex>
  );
};

export default Recipe;
