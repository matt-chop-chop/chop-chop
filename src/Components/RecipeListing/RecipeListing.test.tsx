import { shallow } from "enzyme";
import React from "react";
import RecipeListing, { RecipeListingProps } from "./RecipeListing";
import { recipe } from "@/data/recipe";

const props: RecipeListingProps = {
  recipe: recipe,
};

describe("RecipeListing", () => {
  it("is exported from RecipeListing", () => {
    expect(RecipeListing).toBeDefined();
  });

  it("renders as expected ", () => {
    const wrapper = shallow(<RecipeListing {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
