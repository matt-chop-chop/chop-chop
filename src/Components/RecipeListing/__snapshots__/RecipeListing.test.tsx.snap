// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeListing renders as expected  1`] = `
<Link
  _hover={
    {
      "textDecoration": "none",
    }
  }
  as={
    {
      "$$typeof": Symbol(react.forward_ref),
      "default": [Circular],
      "render": [Function],
    }
  }
  href="/recipe/52922"
>
  <Flex
    _hover={
      {
        "border": "2px solid var(--chakra-colors-light-emphasis)",
        "cursor": "pointer",
        "transform": "scale(1.01)",
        "transition": "transform 250ms ease-in",
      }
    }
    background="var(--chakra-colors-light-background)"
    border="2px solid var(--chakra-colors-light-text)"
    borderRadius={4}
    flexDirection="column"
    height="100%"
    justifyContent="space-between"
    minHeight="300px"
    p={4}
    transition="transform 250ms ease-out"
    width="100%"
  >
    <Box>
      <RecipeImage
        defaultLogoWidth={
          [
            "60%",
            "70%",
            "55%",
          ]
        }
        image="https://www.themealdb.com/images/media/meals/rtwwvv1511799504.jpg/preview"
        maxHeight={
          [
            "200px",
            "200px",
            "200px",
          ]
        }
      />
      <Heading
        as="h3"
        mt={4}
        textDecoration="underline"
        variant="heading-small"
      >
        Prawn & Fennel Bisque
      </Heading>
      <RecipeDetailsGrid
        area="French"
        category="Side"
        gridTemplateColumns={
          [
            "repeat(2, 1fr)",
            "repeat(2, 1fr)",
            "repeat(2, 1fr)",
          ]
        }
        ingredientCount={12}
        time={75}
      />
    </Box>
    <RecipeTags
      tags={
        [
          "Soup",
          "Warm",
          "Seafood",
          "Shellfish",
        ]
      }
    />
  </Flex>
</Link>
`;
